### ISSUE TITLE
VPC name is hardcoded (AWS cloud provider)

### ISSUE THREAD

**Issue by antoineco** [kubernetes/kubernetes#9801]
The name of the VPC used to create ELB instances is [hardcoded inside the AWS cloud provider](https://github.com/GoogleCloudPlatform/kubernetes/blob/release-0.19/pkg/cloudprovider/aws/aws.go#L1372). This prevents the creation of services with type `LoadBalancer` inside a VPC with a different name from what is expected ("kubernetes-vpc" for instance).

A solution could be to query the metadata service in order to get the current VPC id, like:
```bash
MAC_ETH0=`curl -s http://169.254.169.254/latest/meta-data/mac`
VPC_ID=`curl -s http://169.254.169.254/latest/meta-data/network/interfaces/macs/$MAC_ETH0/vpc-id`

echo $VPC_ID
```

**Comment by justinsb** [comment]
Good suggestion.  I thought this was a major screw-up on my part, but it turns out it is just the name (which is only a tag), and we always tag it with the same name in cluster/aws/util.sh.

It is limiting, and getting the VPC ID from the metadata service is a better idea.

Is this actually causing you any problems at the moment?  It is not my best code, but I don't know if we need to change it as the highest priority (?)

**Comment by antoineco** [comment]
Hey @justinsb, I wouldn't say this is a critical issue, it's just confusing for users like me who run in a VPC with a different name since logs do not mention what is actually expected in case of failure.

```
Jun 15 18:02:55 ip-10-0-21-132.eu-west-1.compute.internal kube-controller-manager[5138]: E0615 18:02:55.450394    5138 servicecontroller.go:167] Failed to process service delta. Retrying: failed to create external load balancer for service default/api: Unable to find VPC
```

I can live with it but I'm quite sure other people out there deploy clusters using custom approaches. As a temporary solution I would at least mention this restriction on the Name tag somewhere (getting started guides? services.md?)

**Comment by justinsb** [comment]
How were you able to run in a different VPC though?  Is there a "secret" way of running k8s in AWS other than using cluster/kube-up?  :-)

**Comment by antoineco** [comment]
I can explain our setup so that you can understand the use case:

Basically we run everything on CoreOS. The master, Auto Scaling Group (for
the minions) and corresponding Security Groups are created via
Cloudformation, and we start all services within the cluster using fleet
(including token and certs generation, etc)

staging and prod environments are running in 2 identical VPCs, called
"staging" and "prod" (I'm not super creative :)) Each environment runs
multiple clusters, 1 per availability zone.

I could rename one VPC to "kubernetes-vpc", but then, with the current
code, I would still have to manage my second VPC manually.

I also noticed that the `controller-manager` kept trying to create the ELBs
indefinitely even after I updatef my services back to type `NodePort`, but
since it's only a side effect I should open a new issue for that.

**Comment by justinsb** [comment]
Ah ... CloudFormation!  @bakins is that still the recommended way to run with CoreOS?

**Comment by justinsb** [comment]
This will be fairly easy to fix though; I think we can just drop the name filter, because the VPC should be tagged with a KuberentesCluster={ClusterId} tag, and we grab that from whatever instance we are running on.  I'll put together a PR!

**Comment by antoineco** [comment]
Would that work with multiple clusters inside the same VPC?

**Comment by justinsb** [comment]
Hmmm, I think you're right.  Right now it each k8s cluster spins up a new VPC, so it is one-to-one.  I'd like to relax that, but it isn't there yet.

**Comment by bakins** [comment]
@justinsb Lots of folks are using cloudformation. I, personally, use Terraform.  We (CoreOS) are writing "official" docs/recommendations/examples for Kubernetes on CoreOS based on what we have seen that works.

**Comment by justinsb** [comment]
@bakins  OK, good to know.  Let me know if there's anything we should do in the AWS cloudprovider for this!

I think this issue is valid - it isn't too hard not to hard code this!

**Comment by delianides** [comment]
@justinsb Any idea why I would still be getting "Unable to find VPC"  even if my VPC is named `kubernetes-vpc`? 

The error I see in kube-controller-manager logs is this:

```
Jul 01 15:37:27 ip-172-16-116-11 kube-controller-manager[23734]: E0701 15:37:27.400468   23734 servicecontroller.go:171] Failed to process service delta. Retrying: failed to create external load balancer for service default/my-app: Unable to find VPC
```

**Comment by antoineco** [comment]
@justinsb today I hit another issue which makes me believe the cloudprovider should really be able to determine in which VPC the cluster is running:

I created a node which happened to have the same IP as another instance in a different VPC, and the API Server could not register it because of a conflict.

**Comment by thockin** [comment]
Let's get  fresh triage on this now that we're 1.0

**Comment by justinsb** [comment]
This is indeed important.

@delianides and @antoineco I think your most recent issues are both because of tagging.  I need to make sure that the CloudFormation scripts are setting tags correctly: #11884

Given that, I really like your suggestion @antoineco of using the metadata service - it works even if the tags are wrong!  I'm going to work on this now.

**Comment by jtblin** [comment]
Is there an ETA for this? We deploy kubernetes on coreos in our own VPC in AWS and this is a blocker as we are getting same error as OP.

**Comment by packetcollision** [comment]
I'd also love to know if this is going to be included in 1.1. It was a blocker for us, but we worked around it by running a separate ASG of CoreOS machines outside of the Kube cluster but on the same overlay network that do manual load-balancing into the cluster using Vulcan and Romulus. The ASG automatically adds machines to an ELB. @jtblin if you only need to do http(s) load-balancing, something similar might work for you too.

**Comment by emaildanwilson** [comment]
I'm hitting this from a kube-up deployment as well when I deployed to my own existing VPC. The VPC and subnet were created ahead of time and tagged appropriately. Install worked without error, ELB volumes work but when trying to use a LoadBalancer type service I get "Unable to find VPC" then the controller manager starts DOSing the aws API. I had to delete the service and force a restart of the controller manager on my master node before API calls to aws stopped. I'll experiment a bit more and try out the code that was merged https://github.com/kubernetes/kubernetes/pull/13467 and see if it fixes this too. I see that there was a fix in 1.0.4 for a similar API abuse problem w/ [security groups](https://github.com/kubernetes/kubernetes/pull/11029) but this is a different area of the aws.go code.  GitVersion:"v1.0.3", GitCommit:"61c6ac5f350253a4dc002aee97b7db7ff01ee4ca"

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/pull/13467 just went in.  Does that solve the problem?

**Comment by jtblin** [comment]
It looks like it does, thanks!

Last blocker that I know of to run kubernetes in our VPC in AWS is #11543

**Comment by justinsb** [comment]
This should be fixed by #13467.  Please comment/reopen if not.
